client/src/components/openlayersMap.tsx

```diff
@@ -0,0 +1,115 @@
+import MapPopup from './mapPopup';
+import * as styles from './openlayersMap.module.scss';
+
+interface IMapWrapperProps {
+  features: Feature<Geometry>[],
+};
+
+// The below adapted from
+// https://taylor.callsen.me/using-openlayers-with-react-functional-components/
+const MapWrapper = ({features}: IMapWrapperProps) => {
+  const [map, setMap] = useState<Map>();
+  const [featuresLayer, setFeaturesLayer] = useState<VectorLayer>();
+  const [selectedFeature, setSelectedFeature] = useState<FeatureLike>();
+  const [currentZoom, setCurrentZoom] = useState<number>(4);
+  const [currentOverlayPosition, setCurrentOverlayPosition] = useState<Coordinate>([]);
+
+  const mapElement = useRef() as
+    React.MutableRefObject<HTMLInputElement>;
+
+  // create state ref that can be accessed in OpenLayers onclick callback function
+  //  https://stackoverflow.com/a/60643670
+  const mapRef = useRef() as React.MutableRefObject<Map>;
+  if (map) {
+    mapRef.current = map;
+  }
+
+  useEffect( () => {
+    const view = new View({
+      center: fromLonLat([-86.502136, 32.4687126]),
+      zoom: 4,
+    });
+
+    // create and add initial vector source layer, to be replaced layer
+    const initialFeaturesLayer = new VectorLayer({
+      source: new VectorSource(),
+    });
+
+    const initialMap = new Map({
+      target: mapElement.current,
+      view: view,
+      controls: [],
+    });
+    const currentZoom = Math.floor(initialMap.getView().getZoom()!);
+
+    initialMap.on('moveend', handleMoveEnd);
+    initialMap.on('click', handleMapClick);
+    setMap(initialMap);
+    setCurrentZoom(currentZoom);
+    setFeaturesLayer(initialFeaturesLayer);
+    olms(initialMap, mapStyle);
+  }, []);
+
+
+  // update map if features prop changes
+  useEffect( () => {
+    if (features.length) { // may be empty on first render
+      // set features to map
+      featuresLayer?.setSource(
+          new VectorSource({
+            features: features,
+          }),
+      );
+      const extent = featuresLayer?.getSource().getExtent();
+      if (extent != null) {
+        // fit map to feature extent (with 100px of padding)
+        map?.getView().fit(extent, {
+          padding: [100, 100, 100, 100],
+        });
+      }
+    }
+  }, [features]);
+
+  const handleMapClick = (event: { pixel: any; }) => {
+    const clickedCoord = mapRef.current.getCoordinateFromPixel(event.pixel);
+
+    mapRef.current.forEachFeatureAtPixel(event.pixel, (feature) => {
+      setSelectedFeature(feature);
+      return true;
+    });
+
+    setCurrentOverlayPosition(clickedCoord);
+  };
+
+  const handleMoveEnd = () => {
+    const newZoom = Math.floor(mapRef.current.getView().getZoom()!);
+    if (currentZoom != newZoom) {
+      setCurrentZoom(newZoom);
+    }
+  };
+
+  return (
+    <>
+      <div ref={mapElement} className={styles.mapContainer}/>
+      {map?
+        <MapPopup selectedFeature={selectedFeature!} map={map!} position={currentOverlayPosition} /> :
```

Comment:
OK. I like the use of `!` here as long as React doesn't care about `undefined` vs `null`.

Reply:
Yeah, there is probably a better way to do this - incidentally this results in a bug, which I had filed [here](https://app.zenhub.com/workspaces/justice40-60993f6e05473d0010ec44e3/issues/usds/justice40-tool/263) -  in hte popup itself, we need the `ref` for the html element, but if we conditionally remove it we lose that ref, so we need some other way to conditionally not render something while keeping a ref to it

Reply:
Maybe just hiding/showing the element using css? In Vue, I'd just add a css class that had `display:none;` then control showing it by making the class reactive (add/remove the classname from the element). Not sure if that would work here.

Reply:
Ah, good call - that would definitely work -- in my latest push just now though I check instead in the `position` function for the existence of `selectedFeature` and key off of that instead, this seems to work!